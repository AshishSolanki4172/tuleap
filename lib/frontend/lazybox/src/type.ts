/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { HTMLTemplateResult, TemplateResult } from "lit/html.js";
import type { GroupCollection, LazyboxItem } from "./items/GroupCollection";
import type { HTMLTemplateStringProcessor } from "./index";

export interface Lazybox {
    setDropdownContent: (groups: GroupCollection) => void;
    resetSelection: () => void;
    setSelection: (selection: ReadonlyArray<LazyboxItem>) => void;
    destroy: () => void;
}

export type LazyboxSearchFieldCallback = (query: string) => void;

export type LazyboxSelectionCallback = (selected_value: unknown | null) => void;

export type LazyboxTemplatingCallback = (
    html: typeof HTMLTemplateStringProcessor,
    item: LazyboxItem
) => HTMLTemplateResult;

export type LazyboxNewItemCallback = () => void;

export type LazyBoxWithNewItemButton = {
    readonly new_item_button_label: string;
    readonly new_item_callback: LazyboxNewItemCallback;
};

type LazyBoxWithoutNewItemButton = {
    readonly new_item_callback?: undefined;
};

export type LazyboxOptions = (LazyBoxWithoutNewItemButton | LazyBoxWithNewItemButton) & {
    readonly placeholder: string;
    readonly search_input_placeholder: string;
    readonly is_multiple: boolean;
    readonly templating_callback: LazyboxTemplatingCallback;
    readonly selection_callback: LazyboxSelectionCallback;
    readonly search_field_callback: LazyboxSearchFieldCallback;
};

export type RenderedItemMap = Map<string, RenderedItem>;

export interface RenderedItem {
    id: string;
    template: TemplateResult;
    value: unknown;
    is_disabled: boolean;
    is_selected: boolean;
    group_id: string;
    element: Element;
}

export interface LazyboxComponent {
    wrapper_element: HTMLElement;
    lazybox_element: Element;
    dropdown_element: HTMLElement;
    selection_element: HTMLElement;
    placeholder_element: Element;
    dropdown_list_element: HTMLElement;
    search_field_element: HTMLInputElement;
}

export interface LazyboxSelectionStateSingle {
    selected_item: RenderedItem;
    selected_value_element: DocumentFragment;
}

export interface LazyboxSelectionStateMultiple {
    selected_items: Map<string, RenderedItem>;
    selected_values_elements: Map<string, Element>;
}

export interface ScrollCoordinates {
    x_position: number;
    y_position: number;
}

export interface ManageSelection {
    processSelection(item: Element): void;
    hasSelection(): boolean;
    updateSelectionAfterDropdownContentChange(): void;
    clearSelection(): void;
    setSelection(selection: ReadonlyArray<RenderedItem>): void;
}
