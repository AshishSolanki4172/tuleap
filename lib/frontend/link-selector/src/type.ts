/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { TemplateResult } from "lit/html.js";
import type { LinkSelectorSearchFieldCallback } from "./events/SearchFieldEventCallbackHandler";

export interface LinkSelector {
    destroy: () => void;
}

export interface LinkSelectorOptions {
    readonly placeholder?: string;
    search_field_callback: LinkSelectorSearchFieldCallback;
}

export type RenderedItemMap = Map<string, RenderedItem>;

export interface RenderedItem {
    id: string;
    template: TemplateResult;
    value: string;
    is_disabled: boolean;
    is_selected: boolean;
    group_id: string;
    element: Element;
    target_option: HTMLOptionElement;
}

export interface LinkSelectorComponent {
    wrapper_element: HTMLElement;
    link_selector_element: Element;
    dropdown_element: HTMLElement;
    selection_element: HTMLElement;
    placeholder_element: Element;
    dropdown_list_element: HTMLElement;
    search_field_element: HTMLInputElement;
}

export interface LinkSelectorSelectionStateSingle {
    selected_item: RenderedItem;
    selected_value_element: DocumentFragment;
}

export interface ScrollCoordinates {
    x_position: number;
    y_position: number;
}
