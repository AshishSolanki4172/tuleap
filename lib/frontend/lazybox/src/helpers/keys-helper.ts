/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

export function isEscapeKey(event: KeyboardEvent): boolean {
    return event.key === "Escape" || event.key === "Esc";
}

export function isEnterKey(event: KeyboardEvent): boolean {
    return event.key === "Enter";
}

export function isArrowDown(event: KeyboardEvent): boolean {
    return event.key === "ArrowDown";
}

export function isArrowUp(event: KeyboardEvent): boolean {
    return event.key === "ArrowUp";
}

export function isTabKey(event: KeyboardEvent): boolean {
    return event.key === "Tab";
}

export function isBackspaceKey(event: KeyboardEvent): boolean {
    return event.key === "Backspace";
}
