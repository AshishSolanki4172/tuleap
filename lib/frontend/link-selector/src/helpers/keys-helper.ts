/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

export function isEscapeKey(event: Event): event is KeyboardEvent {
    const keyboard_event = getKeyboardEvent(event);

    return keyboard_event.key === "Escape" || keyboard_event.key === "Esc";
}

export function isEnterKey(event: Event): event is KeyboardEvent {
    const keyboard_event = getKeyboardEvent(event);

    return keyboard_event.key === "Enter";
}

export function isArrowDown(event: Event): event is KeyboardEvent {
    const keyboard_event = getKeyboardEvent(event);

    return keyboard_event.key === "ArrowDown";
}

export function isArrowUp(event: Event): event is KeyboardEvent {
    const keyboard_event = getKeyboardEvent(event);

    return keyboard_event.key === "ArrowUp";
}

export function isTabKey(event: Event): event is KeyboardEvent {
    const keyboard_event = getKeyboardEvent(event);

    return keyboard_event.key === "Tab";
}

export function isShiftKey(event: Event): event is KeyboardEvent {
    const keyboard_event = getKeyboardEvent(event);

    return keyboard_event.key === "Shift";
}

function getKeyboardEvent(event: Event): KeyboardEvent {
    if (!(event instanceof KeyboardEvent)) {
        throw new Error("Event is not a keyboard event");
    }

    return event;
}
