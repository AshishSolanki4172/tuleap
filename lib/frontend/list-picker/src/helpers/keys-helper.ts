/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

export function isEscapeKey(event: KeyboardEvent): boolean {
    return event.key === "Escape" || event.key === "Esc" || event.keyCode === 27;
}

export function isBackspaceKey(event: KeyboardEvent): boolean {
    return event.key === "Backspace" || event.keyCode === 8;
}

export function isEnterKey(event: KeyboardEvent): boolean {
    return event.key === "Enter" || event.keyCode === 13;
}

export function isArrowDown(event: KeyboardEvent): boolean {
    return event.key === "ArrowDown" || event.keyCode === 40;
}

export function isArrowUp(event: KeyboardEvent): boolean {
    return event.key === "ArrowUp" || event.keyCode === 38;
}

export function isTabKey(event: KeyboardEvent): boolean {
    return event.key === "Tab" || event.keyCode === 9;
}

export function isShiftKey(event: KeyboardEvent): boolean {
    return event.key === "Shift" || event.keyCode === 16;
}
