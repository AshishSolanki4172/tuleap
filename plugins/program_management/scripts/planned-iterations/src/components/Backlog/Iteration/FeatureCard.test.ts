/*
 * Copyright (c) Enalean, 2021 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { shallowMount } from "@vue/test-utils";
import { createPlanIterationsLocalVue } from "../../../helpers/local-vue-for-test";

import FeatureCard from "./FeatureCard.vue";

import type { Wrapper } from "@vue/test-utils";

describe("FeatureCard", () => {
    async function getWrapper(): Promise<Wrapper<FeatureCard>> {
        return shallowMount(FeatureCard, {
            localVue: await createPlanIterationsLocalVue(),
            propsData: {
                feature: {
                    background_color: "peggy-pink",
                    has_user_story_planned: false,
                    has_user_story_linked: false,
                    is_open: true,
                    id: 101,
                    uri: "/uri/of/feature",
                    xref: "feature #101",
                    title: "Feature 101",
                    tracker: {
                        color_name: "red-wine",
                    },
                    project: {
                        id: 101,
                        uri: "uri/to/g-pig",
                        label: "Guinea Pigs",
                        icon: "🐹",
                    },
                },
            },
        });
    }

    it("should display the feature as a card with tracker and background color", async () => {
        const wrapper = await getWrapper();
        expect(wrapper.element).toMatchSnapshot();
    });
});
