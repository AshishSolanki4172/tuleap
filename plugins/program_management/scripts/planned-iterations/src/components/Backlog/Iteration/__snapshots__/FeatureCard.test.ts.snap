// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureCard should display the feature as a card with tracker and background color 1`] = `
<div
  class="element-backlog-item"
>
  <div
    class="element-card element-card-red-wine element-card-background-peggy-pink"
    data-test="feature-card"
  >
    <div
      class="element-card-content"
    >
      <div
        class="element-card-xref-label"
      >
        <div
          class="element-card-xref-project"
        >
          <span
            class="element-card-xref-project-icon"
          >
            
                        🐹
                    
          </span>
          
                    Guinea Pigs
                    
          <i
            class="fas fa-long-arrow-alt-right element-card-xref-icon"
          />
        </div>
         
        <a
          class="element-card-xref element-card-xref-red-wine"
          href="/plugins/tracker/?aid=101"
        >
          
                    feature #101
                
        </a>
      </div>
       
      <span
        class="element-card-label"
      >
        Feature 101
      </span>
    </div>
  </div>
</div>
`;
